python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py

```diff
@@ -151,6 +156,30 @@ def only_partition_keys(self, partition_keys: AbstractSet[PartitionKey]) -> "Ass
             & self._compatible_subset,
         )
 
+    class MultiDimInfo(NamedTuple):
+        tw_dim: PartitionDimensionDefinition
+        secondary_dim: PartitionDimensionDefinition
+
+        @property
+        def tw_partition_def(self) -> TimeWindowPartitionsDefinition:
+            check.inst(self.tw_dim.partitions_def, TimeWindowPartitionsDefinition)
+            assert isinstance(
+                self.tw_dim.partitions_def, TimeWindowPartitionsDefinition
+            )  # appease pyright
```

Comment:
[1] double runtime check? Recommend just `assert` since it satisfies runtime and static

Reply:
not a fan of asserts in non-test code -- we can just do:

```
@property
def tw_partition_def(self) -> TimeWindowPartitionsDefinition:
    return check.inst(self.tw_dim.partitions_def, TimeWindowPartitionsDefinition)
```

as check.inst returns a value with the type of the second argument.

Reply:
Unfortunately `check.inst` returns a value with the type of the first arg, which is why we need `assert` for pyright: https://github.com/dagster-io/dagster/blob/sean/asset-partition-range-events/python_modules/dagster/dagster/_check/__init__.py#L621-L628

This has been discussed in the past: https://github.com/dagster-io/internal/discussions/4827

Reply:
I do a double runtime check because like Owen I don't think asserts should test business logic (users can turn them off).

I don't think writing a `check` and an `assert` is a big deal and I plan on continuing the practice.

Reply:
ah my bad I guess I got it mixed up with the opt_inst behavior of narrowing types

if we never expect the assert to fire, is there a reason not to just make it a `cast()`?

not something I have strong opinions on and it's easy to change, so not blocking by any means

Reply:
Yeah can cast I suppose. I guess I like avoiding the import 🤷‍♂️